### PR TITLE
Feature/con 108

### PR DESCRIPTION
**Pull request by samuelraeburn**
Support for getting matched entities and waiting for matching to occur

### CODE REVIEW

File: rticonnextdds_connector/rticonnextdds_connector.py [hunk 1]
```diff
@@ -1000,6 +1042,32 @@ def wait(self, timeout=None):
...
 		_check_retcode(retcode)
 
+	def wait_for_match(self, timeout = None):
+		"""Waits until this output has a matched input.
+
+		If the operation times out, it will raise :class:`TimeoutError`.
+
+		:param number timeout: The maximum time to wait in milliseconds. By default, infinite.
+
+		:return: The change in the current number of matched outputs. If a positive number is returned, the input has matched with new publishers. If a negative number is returned the input has unmatched from an output. It is possible for multiple matches and/or unmatches to be returned (e.g., 0 could be returned, indicating that the input matched the same number of writers as it unmatched).
```
Comment (alexcamposruiz) [inline rticonnextdds_connector/rticonnextdds_connector.py]: matched subscriptions

File: rticonnextdds_connector/rticonnextdds_connector.py [hunk 2]
```diff
@@ -1000,6 +1042,32 @@ def wait(self, timeout=None):
 		retcode = connector_binding.waitForAcknowledgments(self.native, timeout)
 		_check_retcode(retcode)
 
+	def wait_for_match(self, timeout = None):
+		"""Waits until this output has a matched input.
```
Comment (alexcamposruiz) [inline rticonnextdds_connector/rticonnextdds_connector.py]: until this output has a matched subscription

File: rticonnextdds_connector/rticonnextdds_connector.py [hunk 3]
```diff
@@ -1000,6 +1042,32 @@ def wait(self, timeout=None):
 		retcode = connector_binding.waitForAcknowledgments(self.native, timeout)
 		_check_retcode(retcode)
 
+	def wait_for_match(self, timeout = None):
```
Comment (alexcamposruiz) [inline rticonnextdds_connector/rticonnextdds_connector.py]: I've changed my mind about the name of this function. I think wait_for_subscriptions is better.

File: rticonnextdds_connector/rticonnextdds_connector.py [hunk 4]
```diff
@@ -668,6 +684,32 @@ def take(self):
...
+
+		"""
+		if timeout is None:
+			timeout = -1
+		current_count_change = ctypes.c_int()
+		retcode = connector_binding.wait_for_matched_output(self.native, timeout, byref(current_count_change))
+		_check_retcode(retcode)
+		return current_count_change.value
+
+	def get_matched_outputs(self):
```
Comment (alexcamposruiz) [inline rticonnextdds_connector/rticonnextdds_connector.py]: get_matched_publications

File: rticonnextdds_connector/rticonnextdds_connector.py [hunk 5]
```diff
@@ -1000,6 +1042,32 @@ def wait(self, timeout=None):
...
+
+		"""
+		if timeout is None:
+			timeout = -1
+		current_count_change = ctypes.c_int()
+		retcode = connector_binding.wait_for_matched_input(self.native, timeout, byref(current_count_change))
+		_check_retcode(retcode)
+		return current_count_change.value
+
+	def get_matched_inputs(self):
```
Comment (alexcamposruiz) [inline rticonnextdds_connector/rticonnextdds_connector.py]: get_matched_subscriptions

File: test/python/test_rticonnextdds_discovery.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+
+  # Even though we are going to use the discovery_writer_only_output, don't use the
+  # pytest fixture as we want to control the deletion of it from within the
+  # test function
+  def test_unmatched_output(self, discovery_reader_only_input):
+    # To begin with, no matching occurs
+    assert None is discovery_reader_only_input.get_matched_outputs()
+
+    # Create the Connector object containing the matching input
+    writer_connector = rti.Connector(
```
Comment (alexcamposruiz) [inline test/python/test_rticonnextdds_discovery.py]: Any don't you use the `discovery_writer_only_connector` fixture?

File: test/python/test_rticonnextdds_discovery.py [hunk 7]
```diff
@@ -0,0 +1,195 @@
+###############################################################################
+# (c) 2019 Copyright, Real-Time Innovations.  All rights reserved.            #
+# No duplications, whole or partial, manual or electronic, may be made        #
+# without express written permission.  Any such copies, or revisions thereof, #
+# must display this notice unaltered.                                         #
+# This code contains trade secrets of Real-Time Innovations, Inc.             #
+###############################################################################
+
+import pytest,time,sys,os
```
Comment (alexcamposruiz) [inline test/python/test_rticonnextdds_discovery.py]: Are you testing a publication/subscription that doesn't set the entity name Qos?

File: test/python/test_rticonnextdds_discovery.py [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+
+  def test_no_matches_output(self, discovery_writer_only_output):
+    # We should not match with any outputs at this point
+    matches = discovery_writer_only_output.get_matched_inputs()
+    assert matches is None
+
+    # We should timeout if we attempt to wait for a match
+    with pytest.raises(rti.TimeoutError) as excinfo:
+      change_in_matches = discovery_writer_only_output.wait_for_match(1)
+      assert change_in_matches == 0
```
Comment (alexcamposruiz) [inline test/python/test_rticonnextdds_discovery.py]: This assert won't run if the test passes (because it throws TimeoutError).

File: test/python/test_rticonnextdds_discovery.py [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+
+    # Both the input and output should match each other (and nothing else)
+    change_in_matches = the_input.wait_for_match(2000)
+    matches = the_input.get_matched_outputs()
+    assert change_in_matches == 1
+    assert isinstance(matches, list)
+    assert {'name': 'MyWriter'} in matches
+    change_in_matches = the_output.wait_for_match(2000)
+    matches = the_output.get_matched_inputs()
+    assert change_in_matches == 1
```
Comment (alexcamposruiz) [inline test/python/test_rticonnextdds_discovery.py]: why is this 1? You just called wait on line 86. The second call to wait without any new subscription should keep waiting.

Comment (samuelraeburn) [reply test/python/test_rticonnextdds_discovery.py]: The second call to wait is on an output so the 1 is referring to the matched input in this case

File: test/python/test_rticonnextdds_discovery.py (hunk 6 above)
Comment (samuelraeburn) [reply test/python/test_rticonnextdds_discovery.py]: I wanted to be able to force entities to unmatch so I could test the APIs in that case. If i use the fixture I couldn't find a neat way of deleting the connector.

File: test/python/test_rticonnextdds_discovery.py (hunk 7 above)
Comment (samuelraeburn) [reply test/python/test_rticonnextdds_discovery.py]: Added a test for this

Comment (samuelraeburn) [reply test/python/test_rticonnextdds_discovery.py]: Following up on this...
I can't test an actual xsi:nil="true" entity name policy since connector_new will fail due to an invalid XML file.
The name tag in  `<data_reader name="the_name" topic_ref="the_topic"/> ` sets the entity_name and has to be present. I will add a test using an empty string for the name.

Comment (samuelraeburn) [reply test/python/test_rticonnextdds_discovery.py]: It should be possible to test this in the binding though

File: docs/input.rst [hunk 10]
```diff
@@ -32,28 +32,28 @@ this ``connector`` (see :ref:`Create a new *Connector*`).
 Matching with an Output
 ~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-The method :meth:`Input.wait_for_match()` can be used to detect when a compatible
-:class:`Output` is matched or unmatched. It returns the change in the number of matched outputs
-since the last time it was called::
+The method :meth:`Input.wait_for_publications()` can be used to detect when a compatible
+DDS publication is matched or unmatched. It returns the change in the number of
+matched outputs since the last time it was called::
```
Comment (alexcamposruiz) [inline docs/input.rst]: matched publications

File: docs/output.rst [hunk 11]
```diff
@@ -62,28 +62,28 @@ See :class:`Instance` and :ref:`Accessing the data` for more information.
 Matching with an Input
 ~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-The method :meth:`Output.wait_for_match()` can be used to detect when a compatible
-:class:`Input` is matched or unmatched. It returns the change in the number of matched inputs
-since the last time it was called::
+The method :meth:`Output.wait_for_subscriptions()` can be used to detect when a compatible
+DDS subscription is matched or unmatched. It returns the change in the number of
+matched inputs since the last time it was called::
```
Comment (alexcamposruiz) [inline docs/output.rst]: matched subscriptions

File: test/python/test_rticonnextdds_discovery.py [hunk 12]
```diff
@@ -177,19 +186,30 @@ def test_unmatched_output(self, discovery_reader_only_input):
...
+  def test_no_entity_names(self, discovery_connector_no_entity_names):
+    the_output = discovery_connector_no_entity_names.get_output("MyPublisher::MyWriter")
+    # Ensure that the entities match
+    change_in_subs = the_output.wait_for_subscriptions(-1)
+    assert change_in_subs == 1
+
+    # Get the entity names from the matched publication and subscriptions
+    matched_subs = the_output.get_matched_subscriptions()
+    # Since the entities used in this test do not set a entity_name QoS, check that no name is present
+    assert matched_subs == [{'name': ''}]
```
Comment (alexcamposruiz) [inline test/python/test_rticonnextdds_discovery.py]: we discussed adding a test where we create non-Connector readers and writers, in which case the pub/sub name may not be set.

File: docs/input.rst [hunk 13]
```diff
@@ -32,28 +32,28 @@ this ``connector`` (see :ref:`Create a new *Connector*`).
 Matching with an Output
```
Comment (alexcamposruiz) [inline docs/input.rst]: Matching with a Publication

File: docs/input.rst [hunk 14]
```diff
@@ -26,9 +26,35 @@ the *subscriber* named *MySubscriber*::
      <data_reader name="MySquareReader" topic_ref="Square" />
    </subscriber>
 
-This *publisher* is defined inside the *domain_participant* selected to create
+This *subscriber* is defined inside the *domain_participant* selected to create
 this ``connector`` (see :ref:`Create a new *Connector*`).
 
+Matching with an Output
+~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (alexcamposruiz) [inline docs/input.rst]: I think sphinx will complain if the ~~~ don't match the number of characters of the title.

File: test/python/test_rticonnextdds_discovery.py (hunk 12 above)
Comment (samuelraeburn) [reply test/python/test_rticonnextdds_discovery.py]: Added.

File: test/python/test_rticonnextdds_discovery.py [hunk 15]
```diff
@@ -203,13 +203,35 @@ def test_unmatched_output(self, discovery_reader_only_input):
...
+    # We call the _createTestScenario API to create a remote matching reader which
+    # will not set any entity name.
+    retcode = rti.connector_binding._createTestScenario(
+      discovery_writer_only_output.connector.native,
+      0, # RTI_Connector_testScenario_createReader
+      discovery_writer_only_output.native)
+    assert retcode == 0
+
+    # Wait to match with the new reader
+    discovery_writer_only_output.wait_for_subscriptions(-1)
```
Comment (alexcamposruiz) [inline test/python/test_rticonnextdds_discovery.py]: 1) To specify an infinite wait, we should allow to call the function with no args, instead of passing -1
2) We should wait for a finite amount of time to not hang the test

File: test/python/test_rticonnextdds_discovery.py [hunk 16]
```diff
@@ -203,13 +203,35 @@ def test_unmatched_output(self, discovery_reader_only_input):
...
+    # Check that we can handle getting entity_name when it is NULL
+    matches = discovery_writer_only_output.get_matched_subscriptions()
+    assert isinstance(matches, list)
+    assert len(matches) == 1
+    assert isinstance(matches[0], dict)
+    assert matches[0]['name'] is None
```
Comment (alexcamposruiz) [inline test/python/test_rticonnextdds_discovery.py]: 👍

Comment (alexcamposruiz) [reply test/python/test_rticonnextdds_discovery.py]: Is this behavior documented? "The list returned will contain the name of each matched publication/subscription. If one of them didn't specify a name, the list will contain a `None` element instead."

Comment (alexcamposruiz) [reply test/python/test_rticonnextdds_discovery.py]: Add this as a `note::` or add it to the function documentation, if the .rst starts looking too complicated.

File: test/python/test_rticonnextdds_discovery.py (hunk 15 above)
Comment (samuelraeburn) [reply test/python/test_rticonnextdds_discovery.py]: 1. is already possible:
```
	def wait(self, timeout=None):
		if timeout is None:
			timeout = -1
```
2. in next commit
